docs/docs/aztec/concepts/accounts/keys.md

```diff
@@ -3,157 +3,123 @@ title: Keys
+However if one wants to implement authorization logic containing signatures (e.g. ECDSA or Shnorr) they will need signing keys. Usually, an account contract will validate a signature of the incoming payload against a known signing public key.
 
-#include_code nullifier /noir-projects/aztec-nr/value-note/src/value_note.nr rust
+This is a snippet of our Schnorr Account contract implementation, which uses Schnorr signatures for authentication:
 
-Typically, `Npk_m` is stored in a note and later on, the note is nullified using the secret app-siloed version (denoted `nsk_app`).
-`nsk_app` is derived by hashing `nsk_m` with the app contract address and it is necessary to present it to compute the nullifier.
-Validity of `nsk_app` is verified by our protocol kernel circuits.
+#include_code is_valid_impl /noir-projects/noir-contracts/contracts/schnorr_account_contract/src/main.nr rust
 
-## Incoming viewing keys
+### Storing signing keys
 
-The public key (denoted `Ivpk`) is used to encrypt a note for a recipient and the corresponding secret key (`ivsk`) is used by the recipient during decryption.
+Since signatures are fully abstracted, how the public key is stored in the contract is abstracted as well and left to the developer of the account contract. Among a few common approaches are storing the key in a private note, in an immutable private note, using shared mutable state, reusing other in-protocol keys, or a separate keystore. Below, we elaborate on these approaches.
 
-## Outgoing viewing keys
+#### Using a private note​
 
-App-siloed versions of outgoing viewing keys are denoted `ovsk_app` and `Ovpk_app`.
-These keys are used to encrypt a note for a note sender which is necessary for reconstructing transaction history from on-chain data.
-For example, during a token transfer, the token contract may dictate that the sender encrypts the note with value with the recipient's `Ivpk`, but also records the transfer with its own `Ovpk_app` for bookkeeping purposes.
-If these keys were not used and a new device would be synched there would be no "direct" information available about notes that a user created for other people.
+Storing the signing public key in a private note makes it accessible from the entrypoint function, which is required to be a private function, and allows for rotating the key when needed. However, keep in mind that reading a private note requires nullifying it to ensure it is up-to-date, so each transaction you send will destroy and recreate the public key so the protocol circuits can be sure that the notes are not stale.
 
-## Tagging keys
+#### Using an immutable private note​
 
-Used to compute tags in a tagging note discovery scheme.
+Similar to using a private note, but using an immutable private note removes the need to nullify the note on every read. This generates no nullifiers and commitments per transaction. However, it does not allow the user to rotate their key if they lose it.
```

Comment:
add the code snippet like you did with is_valid_impl

Reply:
are you thinking to add a code snippet about private immutable, like [this](https://github.com/AztecProtocol/aztec-packages/blob/d922a13f83e12a942ec66a673f6785d28481d46c/noir-projects/noir-contracts/contracts/schnorr_account_contract/src/main.nr#L28)? or like the public key note like [this](https://github.com/AztecProtocol/aztec-packages/blob/d922a13f83e12a942ec66a673f6785d28481d46c/noir-projects/noir-contracts/contracts/schnorr_account_contract/src/public_key_note.nr#L15-L19)?